server/src/main/java/org/elasticsearch/index/mapper/MappingParserContext.java

```diff
@@ -50,7 +53,8 @@ public MappingParserContext(
         ScriptCompiler scriptCompiler,
         IndexAnalyzers indexAnalyzers,
         IndexSettings indexSettings,
-        IdFieldMapper idFieldMapper
+        IdFieldMapper idFieldMapper,
+        Function<Query, BitSetProducer> bitSetProducer
```

Comment:
It'd be nice if we can avoid this, but I don't see a better way to pass this to the mapper builder. Maybe @martijnvg has a better idea.

Reply:
The `bitSetProducer` is really required for determining what child Lucene docs belong to the root lucene doc. I'm not sure whether there is a cleaner of way of passing this around.